File: uri/estructuras/bora_bora/1439.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Carta {
    int rango;
    char palo;
    
    int valorPalo() const {
        if (palo == 'C') return 0;
        if (palo == 'D') return 1;
        if (palo == 'H') return 2;
        return 3;
    }
    
    bool operator<(const Carta& o) const {
        if (rango != o.rango) return rango < o.rango;
        return valorPalo() < o.valorPalo();
    }
    
    bool coincide(const Carta& o) const {
        return rango == o.rango || palo == o.palo;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int p, m, n;
    while (cin >> p >> m >> n && (p || m || n)) {
        vector<Carta> deck(n);
        for (int i = 0; i < n; i++) {
            cin >> deck[i].rango >> deck[i].palo;
        }
        
        vector<vector<Carta>> manos(p);
        int idx = 0;
        for (int i = 0; i < p; i++) {
            for (int j = 0; j < m; j++) {
                manos[i].push_back(deck[idx++]);
            }
        }
        
        Carta top = deck[idx++];
        vector<Carta> stock;
        for (int i = idx; i < n; i++) {
            stock.push_back(deck[i]);
        }
        reverse(stock.begin(), stock.end());
        
        int dir = 1;
        int jugador = 0;
        
        if (top.rango == 12) {
            dir = -dir;
        }
        
        int penalidad = 0;
        bool skipTurno = false;
        if (top.rango == 7) {
            penalidad = 2;
            skipTurno = true;
        } else if (top.rango == 1) {
            penalidad = 1;
            skipTurno = true;
        } else if (top.rango == 11) {
            skipTurno = true;
        }
        
        while (true) {
            for (int i = 0; i < penalidad && !stock.empty(); i++) {
                manos[jugador].push_back(stock.back());
                stock.pop_back();
            }
            penalidad = 0;
            
            if (skipTurno) {
                skipTurno = false;
                jugador = (jugador + dir + p) % p;
                continue;
            }
            
            int mejor = -1;
            for (int i = 0; i < manos[jugador].size(); i++) {
                if (manos[jugador][i].coincide(top)) {
                    if (mejor == -1 || manos[jugador][mejor] < manos[jugador][i]) {
                        mejor = i;
                    }
                }
            }
            
            if (mejor == -1 && !stock.empty()) {
                Carta robada = stock.back();
                stock.pop_back();
                manos[jugador].push_back(robada);
                if (robada.coincide(top)) {
                    mejor = manos[jugador].size() - 1;
                }
            }
            
            if (mejor != -1) {
                top = manos[jugador][mejor];
                manos[jugador].erase(manos[jugador].begin() + mejor);
                
                if (manos[jugador].empty()) {
                    cout << jugador + 1 << '\n';
                    break;
                }
            
                if (top.rango == 12) {
                    dir = -dir;
                } else if (top.rango == 7) {
                    penalidad = 2;
                    skipTurno = true;
                } else if (top.rango == 1) {
                    penalidad = 1;
                    skipTurno = true;
                } else if (top.rango == 11) {
                    skipTurno = true;
                }
            }
            
            jugador = (jugador + dir + p) % p;
        }
    }
    
    return 0;
}

File: uri/estructuras/bora_bora/1439.rs
use std::io::{self, BufRead, Write, BufWriter};

#[derive(Clone, Copy)]
struct Carta {
    rango: i32,
    palo: char,
}

impl Carta {
    fn valor_palo(&self) -> i32 {
        match self.palo {
            'C' => 0,
            'D' => 1,
            'H' => 2,
            _ => 3,
        }
    }
    
    fn valor(&self) -> i32 {
        self.rango * 10 + self.valor_palo()
    }
    
    fn coincide(&self, otra: &Carta) -> bool {
        self.rango == otra.rango || self.palo == otra.palo
    }
}

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut lines = stdin.lock().lines();
    
    loop {
        let line = lines.next().unwrap().unwrap();
        let nums: Vec<i32> = line.split_whitespace().map(|x| x.parse().unwrap()).collect();
        let p = nums[0] as usize;
        let m = nums[1] as usize;
        let n = nums[2] as usize;
        
        if p == 0 && m == 0 && n == 0 {
            break;
        }
        
        let mut deck: Vec<Carta> = Vec::with_capacity(n);
        for _ in 0..n {
            let line = lines.next().unwrap().unwrap();
            let parts: Vec<&str> = line.split_whitespace().collect();
            let rango: i32 = parts[0].parse().unwrap();
            let palo = parts[1].chars().next().unwrap();
            deck.push(Carta { rango, palo });
        }
        
        let mut manos: Vec<Vec<Carta>> = vec![vec![]; p];
        let mut idx = 0;
        for i in 0..p {
            for _ in 0..m {
                manos[i].push(deck[idx]);
                idx += 1;
            }
        }
        
        let mut top = deck[idx];
        idx += 1;
        let mut stock: Vec<Carta> = deck[idx..].iter().rev().cloned().collect();
        
        let mut dir: i32 = 1;
        let mut jugador: i32 = 0;
        
        if top.rango == 12 {
            dir = -dir;
        }
        
        let mut penalidad = 0;
        let mut skip_turno = false;
        if top.rango == 7 {
            penalidad = 2;
            skip_turno = true;
        } else if top.rango == 1 {
            penalidad = 1;
            skip_turno = true;
        } else if top.rango == 11 {
            skip_turno = true;
        }
        
        loop {
            let j = jugador as usize;
            
            for _ in 0..penalidad {
                if let Some(c) = stock.pop() {
                    manos[j].push(c);
                }
            }
            penalidad = 0;
            
            if skip_turno {
                skip_turno = false;
                jugador = (jugador + dir + p as i32) % p as i32;
                continue;
            }
            
            let mut mejor: Option<usize> = None;
            for i in 0..manos[j].len() {
                if manos[j][i].coincide(&top) {
                    if mejor.is_none() || manos[j][mejor.unwrap()].valor() < manos[j][i].valor() {
                        mejor = Some(i);
                    }
                }
            }
            
            if mejor.is_none() {
                if let Some(robada) = stock.pop() {
                    manos[j].push(robada);
                    if robada.coincide(&top) {
                        mejor = Some(manos[j].len() - 1);
                    }
                }
            }
            
            if let Some(idx) = mejor {
                top = manos[j][idx];
                manos[j].remove(idx);
                
                if manos[j].is_empty() {
                    writeln!(out, "{}", jugador + 1).unwrap();
                    break;
                }
                
                if top.rango == 12 {
                    dir = -dir;
                } else if top.rango == 7 {
                    penalidad = 2;
                    skip_turno = true;
                } else if top.rango == 1 {
                    penalidad = 1;
                    skip_turno = true;
                } else if top.rango == 11 {
                    skip_turno = true;
                }
            }
            
            jugador = (jugador + dir + p as i32) % p as i32;
        }
    }
}
